### PR TITLE
added windows module win_power_plan with integration test

### PR DESCRIPTION
**Pull request by nwsparks**
##### SUMMARY

Added module for adjusting the power plan in Windows systems.

By default Windows sets it to Balanced which will throttle CPU. This module will default to changing to High Performance which will remove the throttling.

It accepts any string value in case people are trying to define custom power plans.

(Apologies if the pr wasn't done right. First time contributing and was unable to find good documentation on the git steps)

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

win_power_plan

##### ANSIBLE VERSION

```
ansible 2.3.2.0
  config file = /opt/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Sep  1 2016, 22:14:00) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by mattclay** [comment]
bot_status

File: lib/ansible/modules/windows/win_power_plan.ps1 [hunk 1]
```diff
@@ -0,0 +1,81 @@
+﻿#!powershell
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.ps1]: We now have a one-line GPL header we use for modules.
```python
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

File: lib/ansible/modules/windows/win_power_plan.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2015, Jon Hawkesworth (@jhawkesworth) <figs@unity.demon.co.uk>
+#
+# This file is part of Ansible
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Same here, add:
```yaml
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

File: lib/ansible/modules/windows/win_power_plan.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = r'''
+---
+module: win_power_plan
+author: "Noah Sparks (@nwsparks)"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: This should be a list:
```yaml
author:
- Noah Sparks (@nwsparks)
```

File: lib/ansible/modules/windows/win_power_plan.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+version_added: "2.3.2"
+short_description: Changes the power plan of a Windows system.
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Should not have a trailing dot.

File: lib/ansible/modules/windows/win_power_plan.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = r'''
+---
+module: win_power_plan
+author: "Noah Sparks (@nwsparks)"
+version_added: "2.3.2"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Should be 2.4 (as we don't backport new modules to older releases).

File: lib/ansible/modules/windows/win_power_plan.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+short_description: Changes the power plan of a Windows system.
+description:
+- win_power_plan will change the power plan of a Windows system to the defined string.
+- Windows defaults to "balanced" which will cause CPU throttling. In some cases it can be preferable
+  to change the mode to "high performance" to increase CPU performance
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Should have a trailing dot. (Only the short_description must not have one)

File: lib/ansible/modules/windows/win_power_plan.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+options:
+  power_plan:
+    description:
+    - String value that indicates the desired power plan. The power plan must already be
+      present on the system. Commonly there will be options for "balanced" and "high performance"
+    - Defaults to "high performance"
+    - Returns all available power plans to stdout in results
+    version_added: "2.3"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Not needed. The module is new in version 2.4.

File: lib/ansible/modules/windows/win_power_plan.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    description: State of the intended power plan
+    returned: success
+    type: boolean
+    sample: True
+all_available_plans:
+    description: source file used for the copy on the target machine
+    returned: always
+    type: array
+    sample: |
+      "all_available_plans": [
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Wouldn't it be easier to return a dictionary with the name as key ?

File: lib/ansible/modules/windows/win_power_plan.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+short_description: Changes the power plan of a Windows system.
+description:
+- win_power_plan will change the power plan of a Windows system to the defined string.
+- Windows defaults to "balanced" which will cause CPU throttling. In some cases it can be preferable
+  to change the mode to "high performance" to increase CPU performance
+options:
+  power_plan:
+    description:
+    - String value that indicates the desired power plan. The power plan must already be
+      present on the system. Commonly there will be options for "balanced" and "high performance"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Trailing dot.

File: lib/ansible/modules/windows/win_power_plan.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    - Defaults to "high performance"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: This should be removed an replaced with:
```yaml
  default: high performance
```

File: lib/ansible/modules/windows/win_power_plan.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    - Returns all available power plans to stdout in results
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: This is not related to the power_plan property. So not relevant here.
(Besides it is incorrect, the available powerplans is not return in the stdout, but as a result value, which is already documented elsewhere)

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
So I think we need to change the module interface, in the sense that:
```yaml
- win_power_plan:
    power_plan: high performance
```
would probably be better as:
```yaml
- win_power_plan:
    name: high performance
```
Since the module manages the power plan, the parameter is the name of that power plan we're working with.

File: lib/ansible/modules/windows/win_power_plan.py (hunk 8 above)
Comment (nwsparks) [reply lib/ansible/modules/windows/win_power_plan.py]: Thanks for the reviews. I'll see if I can convert it. Was just converting the powershell object to json and this was the result, should be able to convert it to a dictionary first though. Will fix documentation errors as well. Also updating the integration test a bit. Will commit again once ready.

Comment (dagwieers) [reply lib/ansible/modules/windows/win_power_plan.py]: If there's no reason not to, I don't mind. But a dictionary with the name as key is easier to use it in Ansible.

Comment (nwsparks) [reply lib/ansible/modules/windows/win_power_plan.py]: It's ok. Not to difficult to accommodate and it makes sense. Pushed changes.

Comment (nwsparks) [reply lib/ansible/modules/windows/win_power_plan.py]: Hmm...I guess I needed to approve these before pushing, sorry.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/32786/10/tests

**Comment by nwsparks** [comment]
Can I choose not to support 2008 non-r2 (Edit: it may actually not be supportable on 2008 non-r2 at all now that I'm looking into it.)? If so how would I define that to make CI skip the tests on it?

Also can anyone help identify why I keep getting this error on the sanity check? I've tried everything I can think of to get it to stop throwing that error. The module is working but this keeps failing even though the line is there and it looks exactly like all the other modules I've checked and the sample..

lib/ansible/modules/windows/win_power_plan.ps1:0:0: E102 Interpreter line is not "#!powershell"

**Comment by nwsparks** [comment]
If I delete all the code and leave just the commented out header in there it passes sanity locally and stops complaining about the interpreter line. I tried taking things out incrementally, but the only thing that resolves it is removing all the uncommented lines. 

It kinda seems like there may be a bug in the sanity check? should I open an issue on this?

**Comment by mattclay** [comment]
@nwsparks The interpreter line error is due to a unicode BOM at the start of the file. Saving the file with ASCII encoding, or UTF8 without a BOM will resolve the issue.

**Comment by nwsparks** [comment]
@mattclay thanks, that was indeed it. I appreciate the help...it was driving me nuts.

**Comment by nwsparks** [comment]
ready_for_review

File: lib/ansible/modules/windows/win_power_plan.py [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: win_power_plan
+short_description: Changes the power plan of a Windows system
+description:
+  - win_power_plan will change the power plan of a Windows system to the defined string.
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Make this `This module wil...`

File: lib/ansible/modules/windows/win_power_plan.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+  - Windows defaults to "balanced" which will cause CPU throttling. In some cases it can be preferable
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: You may want to use C(balanced) and C(high performance) to make it stand out more in the documentation.

File: lib/ansible/modules/windows/win_power_plan.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    to change the mode to "high performance" to increase CPU performance.
+version_added: "2.4"
+author:
+  - Noah Sparks (@nwsparks)
+options:
+  name:
+    description:
+      - String value that indicates the desired power plan. The power plan must already be
+        present on the system. Commonly there will be options for "balanced" and "high performance".
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Again, use C(balanced) and C(high performance) here.

File: lib/ansible/modules/windows/win_power_plan.py [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    required: false
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.py]: Remove this line. By default parameters are not required, unless stated so.

File: lib/ansible/modules/windows/win_power_plan.ps1 [hunk 16]
```diff
@@ -0,0 +1,83 @@
+#!powershell
+# Copyright (c) 2017 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#Requires -Module Ansible.ModuleUtils.Legacy
+
+$params = Parse-Args -arguments $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+$diff_mode = Get-AnsibleParam -obj $params -name "_ansible_diff" -type "bool" -default $false
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.ps1]: So we used `$diff_support` as the default variable for this. I know someone recently started to do it differently, but all other modules do use **diff_support**.

File: lib/ansible/modules/windows/win_power_plan.ps1 [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+    power_plan_name = $name
+    power_plan_enabled = (Get-PowerPlans $name).isactive
+    all_available_plans = Get-PowerPlans
+}
+
+$all_available_plans = Get-PowerPlans
+
+#Terminate if plan is not found on the system
+If (! ($all_available_plans.ContainsKey($name)) ) {
+    Fail-Json $Result "Defined power_plan: ($name) is not available"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.ps1]: Casing of $result must be consistent everywhere.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_power_plan.ps1 (hunk 16 above)
Comment (nwsparks) [reply lib/ansible/modules/windows/win_power_plan.ps1]: Ok, I'll change it. I'm wondering though if I should leave this in and if so, do I need to do any kind of handling for it if my module doesn't do anything for diff mode?

Comment (dagwieers) [reply lib/ansible/modules/windows/win_power_plan.ps1]: Ah, if you don't have diff support, please remove it or we might think it does :-D

Comment (nwsparks) [reply lib/ansible/modules/windows/win_power_plan.ps1]: ok thanks

File: lib/ansible/modules/windows/win_power_plan.ps1 [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+$all_available_plans = Get-PowerPlans
+
+#Terminate if plan is not found on the system
+If (! ($all_available_plans.ContainsKey($name)) ) {
+    Fail-Json $result "Defined power_plan: ($name) is not available"    
+}
+
+If ($check_mode)
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.ps1]: It is generally preferred to put the $check_mode conditionals in the normal decision tree. So enclosing the actual actions that would induce a change. It makes it easier to read, and ensures that at least the return codes and return values are in sync with a normal run.

File: lib/ansible/modules/windows/win_power_plan.ps1 [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+Else
+{
+    Try {
+        $Null = Invoke-CimMethod -InputObject (Get-PowerPlans $name) -MethodName Activate -ea Stop
+        $result.changed = $true
+        $result.power_plan_enabled = (Get-PowerPlans $name).isactive
+        $result.all_available_plans = Get-PowerPlans
+        Exit-Json $result
+    }
+    Catch {
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_power_plan.ps1]: Important here is to only add try-blocks around the statements you expect to fail. If more than one statements needs a try-block, you need more than one block. This way errors can be more fine-grained and it becomes easier to understand what statement caused an exception.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_power_plan.ps1 (hunk 19 above)
Comment (nwsparks) [reply lib/ansible/modules/windows/win_power_plan.ps1]: thanks for the reviews. updated for both changes.

**Comment by nwsparks** [comment]
ready_for_review

**Comment by jborean93** [comment]
Thanks @nwsparks, the only change I would request is to remove the default for name and fail if it is not specified. Are you able to make that change and I'll merge this in before the freeze for 2.4

**Comment by nwsparks** [comment]
@jborean93  thanks. change made. ready_for_review

**Comment by jborean93** [comment]
@nwsparks, thanks for the fixes I had to make a tiny change on the docs after `name` is now required but once the tests pass I will merge it in.

**Review by jborean93 — APPROVED** [review]

**Comment by nwsparks** [comment]
@jborean93 awesome! thanks

**Comment by jborean93** [comment]
@nwsparks, thanks for the PR. When you have time can you add this module to the changelog under the Windows section for 2.4?
